### ISSUE TITLE
How to fix errors in common.ftl

### ISSUE THREAD

**Issue by Ann2827** [InseeFrLab/keycloakify#58]
Hi!
I managed to fix the template. You can look at the implementation here:

- https://github.com/Ann2827/keycloakify/blob/main/src/bin/build-keycloak-theme/generateFtl/common.ftl
- https://github.com/Ann2827/keycloakify/blob/4f7daf3fc5b33281bec2b157e237a1abd028a745/src/bin/build-keycloak-theme/generateFtl/generateFtl.ts#L165

**Errors with "The type of the containing value was: extended_hash+string"**

<img width="1414" alt="Снимок экрана 2021-12-01 в 12 05 39" src="https://user-images.githubusercontent.com/32645809/145021150-b5b0ebf6-782b-4950-99c5-d80313baca71.png">

1. Long loading of the template and a large number of errors in the console can be fixed:

```
<#attempt>
  <#local value = object[key]!"error_data_is_undefined">
<#recover>
  /* couldn't dereference ${key} of object */
  <#continue>
</#attempt>
```
The freemarker documentation advises:
```
<#local value = object[key]!"">
```
But keycloakify app pages expect undefined.
<img width="596" alt="Снимок экрана 2021-12-03 в 16 51 59" src="https://user-images.githubusercontent.com/32645809/145023761-d188478d-f987-4ec3-b792-4d4f71f6dcb3.png">

And I followed a simple path. (So that these changes do not break other applications.)
```
JSON.parse(JSON.stringify(<@objectToJson_please_ignore_errors object=.data_model depth=0 arr=[] ftl="ftl_template_for_replacement" />), (key, value) => value === "error_data_is_undefined" ? undefined : value)
```
After that, 4 errors will remain in the console for the login.ftl page. Keys: loginUpdateTotpUrl, loginUsernameReminderUrl, loginUpdateProfileUrl, loginUpdatePasswordUrl from url object. And the template will work quickly.

2. The following solution is not perfect, but it helps to fix errors. I've looked through all the templates from the base theme. And collected a list of all the urls that frontend uses: "loginAction", "resourcesPath", "resourcesCommonPath", "loginRestartFlowUrl", "loginUrl", "loginResetCredentialsUrl", "registrationUrl", "registrationAction", "oauth2DeviceVerificationAction", "oauthAction", "loginResetCredentialsUrl". (In my fork, I extend the current keycloakify templates with the basic ones. That's why I'm collecting lists for all the templates. This also applies to fieldNames in messagesPerField.)I added the arr=[] prop to each objectToJson_please_ignore_errors. And before:
```
"${key}": <@objectToJson_please_ignore_errors object=value depth=depth+1 arr=arr_keys ftl=ftl />
```
I'm doing the checks:
```
<#local arr_keys = []>
<#if depth==0 && key=="url" >
  <#local arr_keys = ["loginAction", "resourcesPath", "resourcesCommonPath", "loginRestartFlowUrl", "loginUrl", "loginResetCredentialsUrl", "registrationUrl", "registrationAction", "oauth2DeviceVerificationAction", "oauthAction", "loginResetCredentialsUrl"]>
</#if>
```

**One new error will appear in the login.tpl template related to the fieldNames enumeration.**
<img width="1419" alt="Снимок экрана 2021-12-06 в 11 42 35" src="https://user-images.githubusercontent.com/32645809/145029018-1a92def5-b29a-4ef9-9322-8ad403b6d575.png">

This error can be fixed by adding additional checks:
```
<#assign attributes = (profile.attributes)![]>
<#attempt>
  <#list attributes as attribute>
    <#if (attribute.name)??>
      <#assign fieldNames += [attribute.name]>
    </#if>
  </#list>
<#recover>
</#attempt>
```

**Error in the login-reset-password.ftl template**

Key: attemptedUsername from auth object. AttemptedUsername is used in login-reset-password.ftl template. If you try to get the value by the key of the attemptUsername in login-reset-password.ftp template (or in login.ftl template), it will not lead to an error. Therefore, ftlName parameter should appear in the conditions.

At the end of generateFtl.ts file:
```
objectKeys(ftlPlaceholders).forEach(
            id => (ftlCode = ftlCode.replace(id, ftlPlaceholders[id]).replace("ftl_template_for_replacement", pageId)),
        );
```

At the end of common.ftl file:
```
Object.deepAssign(
        out,
        //Removing all the undefined
        JSON.parse(JSON.stringify(<@objectToJson_please_ignore_errors object=.data_model depth=0 arr=[] ftl="ftl_template_for_replacement" />), (key, value) => value === "error_data_is_undefined" ? undefined : value)
    );
```

Before:
```
"${key}": <@objectToJson_please_ignore_errors object=value depth=depth+1 arr=arr_keys ftl=ftl />,
```
```
<#if depth==0 && key=="auth" && ftl=="login-reset-password.ftl" >
  <#local arr_keys = ["showUsername", "showResetCredentials", "showTryAnotherWayLink", "attemptedUsername"]>
<#elseif depth==0 && key=="auth">
  <#local arr_keys = ["showUsername", "showResetCredentials", "showTryAnotherWayLink"]>
</#if>
```

**Error in the login-update-profile.ftl template**

I'm not sure if this page worked before my changes. On my computer, when I tried to open this page for 10 minutes, errors appeared in the console. As a result, the page did not load. After my changes, it became clear that this page is broken. Because the download has become faster. There is an error in the browser window 500 (internal server error).

I got all the top-level keys from the `.data_model` on login-update-profile page: msg, execution, updateProfileCtx, social, auth, requiredActionUrl, message, locale, authenticatorConfigured, login, url, advancedMsg, messagesPerField, client, realm, kcSanitize, scripts, user, properties.

First I excluded all the keys (before `"${key}": <@objectToJson_please_ignore_errors...`).

```
<#if depth==0 && key=="url" ><#continue></#if>
<#if depth==0 && key=="auth" ><#continue></#if>
<#if depth==0 && key=="locale" ><#continue></#if>
<#if depth==0 && key=="realm" ><#continue></#if>
<#if depth==0 && key=="scripts" ><#continue></#if>
<#if depth==0 && key=="message" ><#continue></#if>
<#if depth==0 && key=="client" ><#continue></#if>
<#if depth==0 && key=="messagesPerField" ><#continue></#if>
<#if depth==0 && key=="user" ><#continue></#if>
<#if depth==0 && key=="requiredActionUrl" ><#continue></#if>
<#if depth==0 && key=="msg" ><#continue></#if>
<#if depth==0 && key=="execution" ><#continue></#if>
<#if depth==0 && key=="updateProfileCtx" ><#continue></#if>
<#if depth==0 && key=="social" ><#continue></#if>
<#if depth==0 && key=="authenticatorConfigured" ><#continue></#if>
<#if depth==0 && key=="login" ><#continue></#if>
<#if depth==0 && key=="advancedMsg" ><#continue></#if>
<#if depth==0 && key=="kcSanitize" ><#continue></#if>
<#if depth==0 && key=="properties" ><#continue></#if>
```

I found a problem in updateProfileCtx key. Since frontend does not use this object, I have excluded it altogether.

```
<#if depth==0 && key=="updateProfileCtx" ><#continue></#if>
```

This object has the following keys: lastName, getClass, setSingleAttribute, getUserProfileContext, getFirstAttribute, getAttributeStream, getUsername, setLastName, setAttribute, hashCode, class, email, isEditUsernameAllowed, getAttributes, setUsername, getLastName, firstName, editUsernameAllowed, setEmail, setFirstName, equals, getEmail, attributes, toString, getFirstName, userProfileContext, username.

The real problem is one or more of them.

**Error in the saml-post-form.ftl template (optional)**

At the moment there are two additional templates in my fork: saml-post-form.ftl, login-config-totp.ftl.
`loginAction` key of url object creates an error. That's why I excluded him:

```
<#if depth==0 && key=="url" && ftl=="saml-post-form.ftl" >
  <#local arr_keys = ["resourcesPath", "resourcesCommonPath", "loginRestartFlowUrl", "loginUrl", "loginResetCredentialsUrl", "registrationUrl", "registrationAction", "oauth2DeviceVerificationAction", "oauthAction", "loginResetCredentialsUrl"]>
<#elseif depth==0 && key=="url">
  <#local arr_keys = ["loginAction", "resourcesPath", "resourcesCommonPath", "loginRestartFlowUrl", "loginUrl", "loginResetCredentialsUrl", "registrationUrl", "registrationAction", "oauth2DeviceVerificationAction", "oauthAction", "loginResetCredentialsUrl"]>
</#if>
```

At this stage, all errors in the templates have been fixed.

**Comment by garronej** [comment]
Hi @Ann2827,  

Thank you for the time and effort you spent trying to fix this.   
I will review attentively what you did later on but I am afraid the only way that I know of to reduce warning would be to make a special case for known value of the `kcContext`.  
It is important, however, that the script keeps working in the general case to be future proof.  

> Long loading of the template

I think the problem of templates taking a long time to load only happens in the test environment, when you have a console logging all the `.ftl` errors in real time. Logging all those errors is what takes time. Once in production everything is pretty smooth.  

By the way I think you are  not running the latest version of the code.  
In the latest version, I have added [a note in the README](https://github.com/InseeFrLab/keycloakify#about-the-errors-related-to-objecttojson-in-keycloak-logs) and renamed the function `objectToJson_please_ignore_errors`.

**Comment by garronej** [comment]
Sorry I haven't read your issue in detail yet. Anyway thank you for your contribution!

**Comment by garronej** [comment]
> I'm not sure if this page worked before my changes. On my computer, when I tried to open this page for 10 minutes, errors appeared in the console.  

You are right! At least in Keycloak 15 this template is very very slow!  

Thank you for your help.

# Note for self:  

![image](https://user-images.githubusercontent.com/6702424/145063697-524348b9-7cbc-40a8-9e77-48e4c7c30f63.png)

I access this page by setting a required action)
![image](https://user-images.githubusercontent.com/6702424/145063871-81e287a7-7c7c-435f-a9fa-ae7c3941cd2f.png)

I also noted that this template is no longer user as long as user profile is enabled:  

![image](https://user-images.githubusercontent.com/6702424/145064017-d14d2704-3ee7-4f33-b5a1-0dce974d4609.png)

When user profile is enabled it's [idp-review-user-profile.ftl](https://github.com/keycloak/keycloak/blob/bfce612641a70e106b20b136431f0e4046b5c37f/themes/src/main/resources/theme/base/login/idp-review-user-profile.ftl) that is used instead.  

And it's not yet implemented so it falls-back to default keycloak theme:  

![image](https://user-images.githubusercontent.com/6702424/145064282-c92e2e42-2acb-4544-bf4d-dd7d7788ad53.png)

**Comment by marziman** [comment]
@Ann2827 thanks for that, we would be happy to get your fixes.

**Comment by Ann2827** [comment]
@garronej Hi! My main branch also contains template extension changes and some changes to work in my project through my fork. I created an additional branch with changes just to fix common.ftl and made a pull request.

**Comment by garronej** [comment]
Fixed in [4.2.14](https://www.npmjs.com/package/keycloakify)
